File: examples/complete/10-matmul-bench/go/big.go
// 10-matmul-bench (Go): same workload as the Spectra twin.
// Naive 32x32 matmul, 20 rounds, loop order i,j,k.
// Checksum total = 55869440.
package main

import (
	"fmt"
	"os"
)

func main() {
	const n = 32
	const iters = 400
	total := 0
	for it := 0; it < iters; it++ {
		round := 0
		c00 := 0
		clast := 0
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				cell := 0
				for k := 0; k < n; k++ {
					cell += (i + k) * (k - j)
				}
				if i == 0 && j == 0 {
					c00 = cell
				}
				if i == 31 && j == 31 {
					clast = cell
				}
				round += cell
			}
		}
		if c00 != 10416 || clast != -20336 || round != 2793472 {
			fmt.Fprintln(os.Stderr, "checksum mismatch in round", it)
			os.Exit(1)
		}
		total += round
	}
	if total != 1117388800 {
		fmt.Fprintln(os.Stderr, "checksum mismatch")
		os.Exit(1)
	}
	fmt.Println("MATMUL ok n=32 sum=1117388800")
}

File: examples/complete/10-matmul-bench/rust/big.rs
// 10-matmul-bench (Rust): same workload as the Spectra twin.
// Naive 32x32 matmul, 20 rounds, loop order i,j,k.
// Checksum total = 55869440.

fn main() {
    const N: i64 = 32;
    const ITERS: i64 = 400;
    let mut total: i64 = 0;
    for _ in 0..ITERS {
        let mut round: i64 = 0;
        let mut c00: i64 = 0;
        let mut clast: i64 = 0;
        for i in 0..N {
            for j in 0..N {
                let mut cell: i64 = 0;
                for k in 0..N {
                    cell += (i + k) * (k - j);
                }
                if i == 0 && j == 0 {
                    c00 = cell;
                }
                if i == 31 && j == 31 {
                    clast = cell;
                }
                round += cell;
            }
        }
        if c00 != 10416 || clast != -20336 || round != 2793472 {
            eprintln!("checksum mismatch in round");
            std::process::exit(1);
        }
        total += round;
    }
    if total != 1117388800 {
        eprintln!("checksum mismatch");
        std::process::exit(1);
    }
    println!("MATMUL ok n=32 sum=1117388800");
}
